synapse/notifier.py

```diff
@@ -274,42 +275,60 @@ def _notify_pending_new_room_events(self, max_room_stream_id: int):
         """
         pending = self.pending_new_room_events
         self.pending_new_room_events = []
+
+        users = set()  # type: Set[Union[str, UserID]]
+        rooms = set()  # type: Set[str]
+
         for room_stream_id, event, extra_users in pending:
             if room_stream_id > max_room_stream_id:
                 self.pending_new_room_events.append(
                     (room_stream_id, event, extra_users)
                 )
             else:
-                self._on_new_room_event(event, room_stream_id, extra_users)
+                if (
+                    event.type == EventTypes.Member
+                    and event.membership == Membership.JOIN
+                ):
+                    self._user_joined_room(event.state_key, event.room_id)
+
+                users.update(extra_users)
+                rooms.add(event.room_id)
+
+        if users or rooms:
+            self.on_new_event(
+                "room_key", max_room_stream_id, users=extra_users, rooms=[event.room_id]
+            )
```

Comment:
The use of `extra_users` and `event` after falling off the end of `pending` is confusing to me are these supposed to be `users` and `rooms`?

If this line is correct, it seems like `users` and `rooms` could just be flags.

Reply:
whoops